### PR TITLE
Clean up testsuite

### LINKED ISSUE #196: Ensure only public entry points are used in tests
ex: we shouldn't call `Stream._init()` in the tests, which is an implementation detail of `cuda.core`. We should only test public APIs.

It is fine to also have tests for internal modules/functionalities, but they should come after we have coverage for public APIs.

> I don't think we have "relational tests" added yet (ex: `Device.create_stream()` returns a `Stream`; `Stream.record()` returns an `Event`, etc), but let's add them in the next PR.

_Originally posted by @leofang in https://github.com/NVIDIA/cuda-python/pull/153#pullrequestreview-2400399182_

### PR DESCRIPTION
**Pull request by ksimpson-work**
Clean up the testsuite by:

1. Removing calls to internal entry-points in favour of public ones. 
2. Updating the use of fixtures to handle context initialization on a per-test basis. This includes a function level fixture to init, and then subsequently de-init when calling a method which requirts context initalization (usually via device.set_current()). It also includes a fixture to deinit after some tests which call set_current() as part of their logic.
3. As an effect of moving to public entry-points, the tests become more relational. The use of examples as additional tests also increases relational test coverage'

closes #196

### CODE REVIEW

File: cuda_core/tests/conftest.py [hunk 1]
```diff
@@ -7,10 +7,23 @@
 # is strictly prohibited.
 
 from cuda.core.experimental._device import Device
```
Comment (leofang) [inline cuda_core/tests/conftest.py]: I think `Device` is public API, no need to access it through a private module:
```suggestion
from cuda.core.experimental import Device
```

File: cuda_core/tests/conftest.py [hunk 2]
```diff
@@ -7,10 +7,23 @@
 # is strictly prohibited.
 
 from cuda.core.experimental._device import Device
+from cuda.core.experimental import _device
+from cuda import cuda
```
Comment (leofang) [inline cuda_core/tests/conftest.py]: nit: future proof
```suggestion
try:
    from cuda.bindings import driver
except ImportError:
    from cuda import cuda as driver
```

File: cuda_core/tests/conftest.py [hunk 3]
```diff
@@ -7,10 +7,23 @@
...
 
-@pytest.fixture(scope="module")
+@pytest.fixture(scope="function")
 def init_cuda():
     device = Device()
     device.set_current()
-    
\ No newline at end of file
+    yield
+    handle_return(cuda.cuCtxPopCurrent())
```
Comment (leofang) [inline cuda_core/tests/conftest.py]:
```suggestion
    handle_return(driver.cuCtxPopCurrent())

```

File: cuda_core/tests/conftest.py [hunk 4]
```diff
@@ -7,10 +7,23 @@
...
+    with _device._tls_lock:
+        del _device._tls.devices
+
+@pytest.fixture(scope="function")
+def deinit_cuda():
+    yield
+    handle_return(cuda.cuCtxPopCurrent())
```
Comment (leofang) [inline cuda_core/tests/conftest.py]:
```suggestion
    handle_return(driver.cuCtxPopCurrent())

```

File: cuda_core/tests/test_device.py [hunk 5]
```diff
@@ -8,10 +8,13 @@
 
 from cuda import cuda, cudart
 from cuda.core.experimental._device import Device
```
Comment (leofang) [inline cuda_core/tests/test_device.py]: ditto
```suggestion
from cuda.core.experimental import Device
```

File: cuda_core/tests/test_event.py [hunk 6]
```diff
@@ -6,34 +6,46 @@
 # this software and related documentation outside the terms of the EULA
 # is strictly prohibited.
 
-from cuda import cuda
-from cuda.core.experimental._event import  EventOptions, Event
-from cuda.core.experimental._utils import handle_return
+from cuda.core.experimental._event import  EventOptions
 from cuda.core.experimental._device import Device
```
Comment (leofang) [inline cuda_core/tests/test_event.py]: ditto
```suggestion
from cuda.core.experimental import Device, EventOptions
```

File: cuda_core/tests/test_event.py [hunk 7]
```diff
@@ -6,34 +6,46 @@
...
     assert event.is_done == True
 
-def test_handle():
-    options = EventOptions()
-    event = Event._init(options)
+def test_handle(init_cuda):
+    options = EventOptions(enable_timing=False)
+    stream = Device().create_stream()
+    event = stream.record(options=options)
     assert isinstance(event.handle, int)
```
Comment (leofang) [inline cuda_core/tests/test_event.py]: nit: I think we should just move the assert to one of the tests. The test body overlaps with other tests.

File: cuda_core/tests/test_event.py [hunk 8]
```diff
@@ -6,34 +6,46 @@
...
 
-from cuda import cuda
-from cuda.core.experimental._event import  EventOptions, Event
-from cuda.core.experimental._utils import handle_return
+from cuda.core.experimental._event import  EventOptions
 from cuda.core.experimental._device import Device
-import pytest
 
-def test_is_timing_disabled():
+def test_is_timing_disabled(init_cuda):
```
Comment (leofang) [inline cuda_core/tests/test_event.py]: Let's call it `test_timing` and parametrize this test with 3 `timing` options: `None`, `True`, `False`. Specialize a case for `None`: We need to test the default (without any `EventOptions`) is no timing (IIRC).

File: cuda_core/tests/test_launcher.py [hunk 9]
```diff
@@ -6,14 +6,12 @@
 # this software and related documentation outside the terms of the EULA
 # is strictly prohibited.
 
-from cuda import cuda
 from cuda.core.experimental._launcher import LaunchConfig
 from cuda.core.experimental._stream import Stream
 from cuda.core.experimental._device import Device
```
Comment (leofang) [inline cuda_core/tests/test_launcher.py]: nit
```suggestion
from cuda.core.experimental import Device, LaunchConfig
from cuda.core.experimental._stream import Stream
```

File: cuda_core/tests/test_program.py [hunk 10]
```diff
@@ -8,7 +8,6 @@
 
 from cuda.core.experimental._program import Program
```
Comment (leofang) [inline cuda_core/tests/test_program.py]: nit
```suggestion
from cuda.core.experimental import Program
```

**Review by leofang — CHANGES_REQUESTED** [review]
Thanks, Keenan, left some nits.

**Comment by ksimpson-work** [comment]
Thanks @leofang for those comments. I addressed all of them and made the changes.

File: cuda_core/tests/conftest.py [hunk 11]
```diff
@@ -5,12 +5,28 @@
...
+    handle_return(driver.cuCtxPopCurrent())
+    with _device._tls_lock:
+        del _device._tls.devices
+
+@pytest.fixture(scope="function")
+def deinit_cuda():
+    yield
+    handle_return(driver.cuCtxPopCurrent())
+    with _device._tls_lock:
+        del _device._tls.devices
```
Comment (leofang) [inline cuda_core/tests/conftest.py]: nit: let's try not to duplicate code
```suggestion
    _device_unset_current()

def _device_unset_current():
    handle_return(driver.cuCtxPopCurrent())
    with _device._tls_lock:
        del _device._tls.devices

@pytest.fixture(scope="function")
def deinit_cuda():
    yield
    _device_unset_current()
```

File: cuda_core/tests/test_device.py [hunk 12]
```diff
@@ -6,12 +6,24 @@
...
-import pytest
+try:
+    from cuda.bindings import driver
+except ImportError:
+    from cuda import cuda as driver
 
+try:
+    from cuda.bindings import runtime
+except ImportError:
+    from cuda import cudart as runtime
```
Comment (leofang) [inline cuda_core/tests/test_device.py]: nit
```suggestion
try:
    from cuda.bindings import driver, runtime
except ImportError:
    from cuda import cuda as driver
    from cuda import cudart as runtime
```

File: cuda_core/tests/test_stream.py [hunk 13]
```diff
@@ -6,68 +6,68 @@
...
     with pytest.raises(TypeError):
         stream.record(event="invalid_event")
 
-def test_stream_wait_event():
-    stream = Stream._init(options=StreamOptions())
+def test_stream_wait_event(init_cuda):
+    stream = Device().create_stream(options=StreamOptions())
     event = Event._init()
     stream.record(event)
     stream.wait(event)  # Should not raise any exceptions
```
Comment (leofang) [inline cuda_core/tests/test_stream.py]: we can actually test the real-world use case (also get rid of `Event._init()`):
```suggestion
def test_stream_wait_event(init_cuda):
    s1 = Device().create_stream()
    s2 = Device().create_stream()
    e1 = s1.record()
    s2.wait(e1)  # Should not raise any exceptions
    s2.sync()
```

**Review by leofang — APPROVED** [review]
Thanks, Keenan! Left 3 nits, feel free to merge after addressing them.

**Review by vzhurba01 — APPROVED** [review]

**Pull request merged**